pex/pex_bootstrapper.py

```diff
@@ -442,57 +442,58 @@ def ensure_venv(pex):
             collisions = []
             for chars in range(8, len(venv_hash) + 1):
                 entropy = venv_hash[:chars]
-                short_venv_path = os.path.join(short_venv_dir, entropy)
-                try:
-                    os.symlink(venv_dir, short_venv_path)
-                    break
-                except OSError as e:
-                    if e.errno != errno.EEXIST:
-                        raise e
-                    collisions.append(short_venv_path)
-                    if entropy == venv_hash:
-                        raise RuntimeError(
-                            "The venv for {pex} at {venv} has hash collisions with {count} other "
-                            "{venvs}!\n{collisions}".format(
+                short_venv_dir = os.path.join(pex_info.pex_root, "venvs", "s", entropy)
```

Comment:
The GitHub diff UI is pretty bad. The only change here is to use a nested `atomic_directory` with `if short_venv.is_finalized; continue` being the new collision detection mechanism that was previously performed by an eager `os.symlink` in a `try` / `except`. To effect this change, instead of the structure `PEX_ROOT/venvs/short/<short hash> -> <real venv>` the structure becomes `PEX_ROOT/venvs/s/<short hash>/venv -> <real venv>`. Notably, 'short/' is replaced with 's/venv/'. SO we only gain 1 character for the short links.